File: backend/src/main/java/com/codeloom/backend/exception/RestExceptionHandler.java
package com.codeloom.backend.exception;

import jakarta.servlet.http.HttpServletRequest;
import java.time.LocalDateTime;
import java.util.stream.Collectors;
import org.slf4j.*;
import org.springframework.http.*;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.core.AuthenticationException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

@RestControllerAdvice
public class RestExceptionHandler {
  private static final Logger log = LoggerFactory.getLogger(RestExceptionHandler.class);

  @ExceptionHandler(Exception.class)
  public ResponseEntity<ErrorResponse> handleException(Exception e, HttpServletRequest r) {
    ErrorResponse x = errorResponse(e, r);
    return ResponseEntity.status(x.status()).body(x);
  }

  @ExceptionHandler({
    AuthenticationException.class,
    AccessDeniedException.class,
    ResponseStatusException.class,
    MethodArgumentNotValidException.class,
    HttpMessageNotReadableException.class
  })
  public ResponseEntity<ErrorResponse> handleKnown(Exception e, HttpServletRequest r) {
    return handleException(e, r);
  }

  private ErrorResponse errorResponse(Exception e, HttpServletRequest r) {
    log.error("Exception from {} {}", r.getMethod(), r.getRequestURI(), e);
    int s = 500;
    String m = "Something went wrong";
    java.util.Map<String, ?> p = null;
    if (e instanceof AuthenticationException) {
      s = 401;
      m = "Authentication required";
    } else if (e instanceof AccessDeniedException) {
      s = 403;
      m = "Access denied";
    } else if (e instanceof ResponseStatusException x) {
      s = x.getStatusCode().value();
      m = x.getMessage();
    } else if (e instanceof HttpMessageNotReadableException) {
      s = 400;
      m = "Body is malformed";
    } else if (e instanceof MethodArgumentNotValidException x) {
      s = x.getStatusCode().value();
      m =
          "Validation failed: "
              + (x.getFieldError() == null
                  ? "Unknown error"
                  : x.getFieldError().getDefaultMessage());
      p =
          x.getFieldErrors().stream()
              .collect(
                  Collectors.toMap(
                      f -> f.getField(),
                      f -> f.getDefaultMessage() == null ? "" : f.getDefaultMessage(),
                      (a, b) -> b));
    }
    return new ErrorResponse(s, m, LocalDateTime.now(), r.getRequestURI(), p);
  }
}

File: backend/src/main/kotlin/com/codeloom/backend/exception/RestExceptionHandler.kt
package com.codeloom.backend.exception

import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.servlet.http.HttpServletRequest
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.core.AuthenticationException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.server.ResponseStatusException
import java.time.LocalDateTime

private val logger = KotlinLogging.logger { }

@RestControllerAdvice
class RestExceptionHandler {
    @ExceptionHandler(AuthenticationException::class)
    fun handleException(
        e: AuthenticationException,
        servlet: HttpServletRequest,
    ): ResponseEntity<ErrorResponse> {
        val error = errorResponse(e, servlet)
        return ResponseEntity.status(error.status).body(error)
    }

    @ExceptionHandler(AccessDeniedException::class)
    fun handleException(
        e: AccessDeniedException,
        servlet: HttpServletRequest,
    ): ResponseEntity<ErrorResponse> {
        val error = errorResponse(e, servlet)
        return ResponseEntity.status(error.status).body(error)
    }

    @ExceptionHandler(ResponseStatusException::class)
    fun handleException(
        e: ResponseStatusException,
        servlet: HttpServletRequest,
    ): ResponseEntity<ErrorResponse> {
        val error = errorResponse(e, servlet)
        return ResponseEntity.status(error.status).body(error)
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleException(
        e: MethodArgumentNotValidException,
        servlet: HttpServletRequest,
    ): ResponseEntity<ErrorResponse> {
        val error = errorResponse(e, servlet)
        return ResponseEntity.status(error.status).body(error)
    }

    @ExceptionHandler(Exception::class)
    fun handleException(
        e: Exception,
        servlet: HttpServletRequest,
    ): ResponseEntity<ErrorResponse> {
        val error = errorResponse(e, servlet)
        return ResponseEntity.status(error.status).body(error)
    }

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handleException(
        e: HttpMessageNotReadableException,
        servlet: HttpServletRequest,
    ): ResponseEntity<ErrorResponse> {
        val error = errorResponse(e, servlet)
        return ResponseEntity.status(error.status).body(error)
    }

    private fun errorResponse(
        e: Exception,
        servlet: HttpServletRequest,
    ): ErrorResponse {
        logger.error(e) {
            "Exception from ${servlet.method} ${servlet.requestURI}"
        }
        var errorResponse = ErrorResponse(
            status = HttpStatus.INTERNAL_SERVER_ERROR.value(),
            message = "Something went wrong",
            timestamp = LocalDateTime.now(),
            path = servlet.requestURI,
        )
        errorResponse = when (e) {
            is AuthenticationException ->
                errorResponse.copy(
                    status = HttpStatus.UNAUTHORIZED.value(),
                    message = "Authentication required",
                )

            is AccessDeniedException ->
                errorResponse.copy(
                    status = HttpStatus.FORBIDDEN.value(),
                    message = "Access denied",
                )

            is ResponseStatusException ->
                errorResponse.copy(
                    status = e.statusCode.value(),
                    message = e.message,
                )

            is HttpMessageNotReadableException ->
                errorResponse.copy(
                    status = HttpStatus.BAD_REQUEST.value(),
                    message = "Body is malformed",
                )

            is MethodArgumentNotValidException -> {
                errorResponse.copy(
                    status = e.statusCode.value(),
                    message = "Validation failed: ${e.fieldError?.defaultMessage ?: "Unknown error"}",
                    payload =
                        e.fieldErrors.associate { fieldError ->
                            fieldError.field to fieldError.defaultMessage
                        },
                )
            }

            else -> errorResponse
        }

        return errorResponse
    }
}
